policy: Enable ingress CIDR-dependent L3 policy (FromCIDR + ToPorts)

christarazi opened the pull request:
See commit msgs.

Updates: https://github.com/cilium/cilium/issues/4129

**Note to backporters**: If there are conflicts with the tests, then test coverage commits do not need to be backported; only enabling the policy.

christarazi committed:
policy: Allow CIDR-dependent L4 support on ingress

This allows policies such as FromCIDR + ToPorts on ingress.

Signed-off-by: Chris Tarazi <chris@isovalent.com>

christarazi committed:
test: Add policy test for ingress FromCIDR+ToPorts

This test validates that the new ingress FromCIDR+ToPorts policy
(CIDR-dependent L4) allows ingress traffic from an allowed IP address
range, to specific ports.

The setup of this test requires a third node (k8s3) which doesn't have
Cilium installed on it by making use of the NO_CILIUM_ON_NODE
environment variable of the test suite.

The test validates the policy by checking that ingress traffic is:
  1) Allowed before any policies
  2) Disallowed by applying a default deny ingress policy
  3) Allowed again after applying "cnp-ingress-from-cidr-to-ports"

Signed-off-by: Chris Tarazi <chris@isovalent.com>

christarazi committed:
test/helpers: Add MonitorEndpointStart

This is useful for monitoring specific endpoints.

Signed-off-by: Chris Tarazi <chris@isovalent.com>

christarazi committed:
test/helpers: Add adding/deleting route wrappers

Signed-off-by: Chris Tarazi <chris@isovalent.com>

christarazi committed:
test: Factor out importing policy function

Signed-off-by: Chris Tarazi <chris@isovalent.com>

joestringer commented on pkg/policy/api/ingress.go:
```diff
@@ -29,13 +29,6 @@ import (
 //   the effects of any Requires field in any rule will apply to all other
 //   rules as well.
-//
-// - For now, combining ToPorts, FromCIDR, and FromEndpoints in the same rule
-//   is not supported and any such rules will be rejected. In the future, this
-//   will be supported and if multiple members of this structure are specified,
-//   then all members must match in order for the rule to take effect. The
-//   exception to this rule is the Requires field, the effects of any Requires
-//   field in any rule will apply to all other rules as well.
```
We can probably simplify this to:

`Combining FromCIDR and FromEndpoints in the same rule is not allowed.`

christarazi replied:
Sure. Where is that enforced? I just tried to apply:

```
apiVersion: "cilium.io/v2"
kind: CiliumNetworkPolicy
description: "Policy to allow traffic external to cluster inwards via CIDR and port"
metadata:
  name: "cnp-ingress-from-cidr-to-ports"
spec:
  endpointSelector:
    matchLabels:
      zgroup: testDS
  ingress:
  - fromEndpoints: 
    - {}
  - toPorts:
    - ports:
      - port: "80"
        protocol: TCP
    fromCIDR:
      - 192.168.36.0/24
      - 192.168.36.13/32
```

And it applied fine. No msgs in the agent logs.

joestringer replied:
It should be done by `sanitize()` in `pkg/policy/api/rule_validation.go`.

Note that the policy above is two rules, one with `fromEndpoints` and the other with `toPorts` + `fromCIDR`. This is valid. The `-` before `toPorts` would need to be removed to test this.

joestringer commented on test/k8sT/Policies.go:
```diff
@@ -1122,6 +1117,170 @@ var _ = Describe("K8sPolicyTest", func() {
+				RedeployCilium(
+					kubectl,
+					ciliumFilename,
+					mergeMap(deepCopyMap(daemonCfg), map[string]string{
```
:thinking: nit: we do similar things elsewhere in the suite where we customize the daemon options. Not sure why only now we introduce the `mergeMap`/`deepCopyMap` logic (or whether we should be using this more comprehensively elsewhere).

christarazi replied:
Yeah I was surprised to learn that we had no standard mechanism for this. At the time I didn't think too much about it, and wrote the above because it was easy enough. 

Now that I take a closer look at the codebase, there aren't a lot of places where we merge configuration over with new values. The majority of occurrences where we reuse the same `map[string]string` are simply to toggle a whole different suite of flags, and rely on the previous flags to revert back to their defaults. Rarely do you see an inheritance so to speak.  

There are a handful of inheritance occurrences, and most of them are in the same suite, `Policies`. There's one occurrence in `DatapathConfiguration`.

Given that, there's not a huge need for  `mergeMap`/ `deepCopyMap` to be used everywhere. However, having a helper function to deploy Cilium with the ability to merge configuration could useful. I'll introduce that in this suite, and see what people think.

christarazi replied:
Implemented in e5f6bd923df820b21c2ad8eb301fefdb0ae9ce40

joestringer commented on test/k8sT/Policies.go:
```diff
@@ -1122,6 +1117,170 @@ var _ = Describe("K8sPolicyTest", func() {
+				RedeployCilium(kubectl, ciliumFilename, daemonCfg)
+			})
+
+			testConnecitivty := func(expectSuccess, expectError bool) {
```
nit:

```suggestion
			testConnectivity := func(expectSuccess, expectError bool) {
```

joestringer reviewed (changes requested):
Nice work, just few minor nits.

christarazi commented on test/k8sT/Policies.go:
```diff
@@ -1122,6 +1117,170 @@ var _ = Describe("K8sPolicyTest", func() {
+					helpers.IPDelRoute(backendPodIP, hostIPOfBackendPod))
+
+				// Revert the Cilium installation that was changed in the
+				// previous Context
```
```suggestion
				// this Context
```

pchaigno commented on pkg/policy/api/rule_validation.go:
```diff
@@ -113,8 +113,8 @@ func (i *IngressRule) sanitize() error {
-		"FromCIDRSet":   false,
+		"FromCIDR":      true,
+		"FromCIDRSet":   true,
 		"FromEntities":  true,
```
If they are now all supported, can't the check on `ToPorts` below be removed?

christarazi replied:
Good catch, removed.

christarazi committed:
test/helpers: Add merge cfg Cilium deploy variant

This helper is useful for situations where you'd want to preserve the
previous Cilium configuration when redeploying. This helper abstracts
away the need to manually create a copy of the new merged configuration
before calling RedeployCilium.

Example usage:

```
daemonCfg := map[string]string{ ... }

(...)

RedeployCiliumWithMerge(..., daemonCfg, map[string]string{
  "flag-previously-set-in-daemonCfg": "1",
  "new-flag-also":                    "abc",
})
```

Signed-off-by: Chris Tarazi <chris@isovalent.com>

pchaigno commented on test/k8sT/Policies.go:
```diff
@@ -1115,6 +1117,191 @@ var _ = Describe("K8sPolicyTest", func() {
+						// https://github.com/cilium/cilium/issues/12141
+						"global.masquerade":   "false",
+						"global.devices":      "",
+						"global.hostFirewall": "disabled",
```
As I understood #12141, only one of `global.masquerade` and `global.hostFirewall` needs to be disabled here, no?

I would recommend disabling the host firewall here since you're running with per-endpoint routes `global.endpointRoutes` and the host firewall doesn't support that currently (currently results in bypass and soon in fatal with #12495).

pchaigno replied:
By the way, why are per-endpoint routes enabled here?

christarazi replied:
Per-endpoint routes are usually used when doing direct routing so I just did that without much thought. I've removed it as it is not explicitly required.

As for #12141, it turns out I can enable masquerading (I think I misinterpreted the aforementioned issue). However, I need to explicity do the following:

```
"global.devices":      "",
"global.hostFirewall": "false",
```

for the tests to pass.

I don't understand why clearing out the `devices` is required, but if I don't clear them, the tests fail by the wrong identity to incoming traffic (sets it to `WORLD`). I think this is what Jarno was seeing in this thread: https://cilium.slack.com/archives/CDKG8NNHK/p1592425114171600.

As for the host firewall, do you think it should be disabled, or that it shouldn't matter? I can see why it's having an impact, but I am not sure if that's ideal.

pchaigno replied:
> I don't understand why clearing out the `devices` is required, but if I don't clear them, the tests fail by the wrong identity to incoming traffic (sets it to `WORLD`). I think this is what Jarno was seeing in this thread: https://cilium.slack.com/archives/CDKG8NNHK/p1592425114171600.

Yep, I think we have a bug in the logic we use to derive the source security ID in bpf_host. That identity is then passed to the bpf_lxc programs (and used in the host firewall). I haven't had time to take a second look yet.

> As for the host firewall, do you think it should be disabled, or that it shouldn't matter? I can see why it's having an impact, but I am not sure if that's ideal.

The host firewall shouldn't matter here. However, it requires `devices` to be set, so we'll have to disable it as long as `devices` aren't set.

> As for #12141, it turns out I can enable masquerading (I think I misinterpreted the aforementioned issue). However, I need to explicity do the following:
> 
> ```
> "global.devices":      "",
> "global.hostFirewall": "false",
> ```
> 
> for the tests to pass.

The other way around should work too: disabling `masquerade` and enabling `hostFirewall`+`devices`. It might be best to keep it that way (i.e., host firewall disabled) as I'll need to add another test anyway to test FromCIDR+ToPorts as a host policy; I will likely test the reverse option in that other test (i.e., masquerading disabled).

pchaigno commented on test/k8sT/Policies.go:
```diff
@@ -1115,6 +1117,191 @@ var _ = Describe("K8sPolicyTest", func() {
+
+				outsideNodeName, outsideIP = kubectl.GetNodeInfo(helpers.GetNodeWithoutCilium())
+				Expect(outsideNodeName).ShouldNot(BeEmpty())
+				Expect(outsideIP).ShouldNot(BeEmpty())
```
I think these two error cases are already checked at: https://github.com/cilium/cilium/blob/c78b1f911a763864bf75e7fa5ccf9e238f483a94/test/helpers/kubectl.go#L966-L967 and https://github.com/cilium/cilium/blob/c78b1f911a763864bf75e7fa5ccf9e238f483a94/test/helpers/kubectl.go#L988-L989

pchaigno commented on test/k8sT/Policies.go:
```diff
@@ -1115,6 +1117,191 @@ var _ = Describe("K8sPolicyTest", func() {
+				hostNodeName = backendPod.Spec.NodeName // Save the name of node backend pod is on
+
+				// Add the route on the outside node to the backend pod IP
+				// directly.
```
Nit: might be useful to clarify here why this is required. If I followed correctly, because we want to ensure the client's IP is preserved at the destination pod?

christarazi replied:
Added more clarification, let me know if it makes more sense now.

pchaigno commented on test/k8sT/Policies.go:
```diff
@@ -1115,6 +1117,191 @@ var _ = Describe("K8sPolicyTest", func() {
+				By("Testing that connectivity from outside node is %s", action)
+
+				// This Eventually waits for the traffic to eventually succeed
+				// or fail (as we wait for the policy to take effect).
```
Doesn't `CiliumPolicyAction` (called by `importPolicy`) already wait for the policy to be applied in all endpoints?

christarazi replied:
Yeah, I think you're right in that the `Eventually` may not be required. I thought that the test was flakey in CI, and I wasn't seeing any flakes locally. 

It turns out the test isn't flakey from what I can tell. It was due to using per-endpoint routes in this test context..The demo pods (test pods) were never deleted and redeployed to run under this configuration. I believe this was causing the tests to fail in CI. I wasn't seeing it locally because I had commented out the top-level deploy of CIlium for the sake of speed, so the only deploy of Cilium happening was in this test's context. In other words, the demo pods' plumbing was in-sync with what Cilium was expecting when I was running locally, but not in CI.

This should be fixed now that I've removed configuring per-endpoint routes.

pchaigno commented on test/k8sT/manifests/cnp-ingress-from-cidr-to-ports.yaml:
```diff
@@ -0,0 +1,17 @@
+        protocol: TCP
+    fromCIDR:
+      - 192.168.36.0/24
+      - 192.168.36.13/32
```
Why are both CIDRs specified here and not only the first?

christarazi replied:
Ah, good catch. I only meant for `192.168.36.13/32` to be in there.

pchaigno reviewed (commented):
A few nits and questions below. LGTM overall!

It looks like implementing the same test for the host firewall is going to be fairly easy now :tada: Thanks!

christarazi commented on test/k8sT/Policies.go:
```diff
@@ -1122,6 +1117,183 @@ var _ = Describe("K8sPolicyTest", func() {
+						"global.autoDirectNodeRoutes": "true",
+
+						// These are required to avoid failing on net-next.
+						"global.masquerade":    "false",
+						"config.bpfMasquerade": "false",
```
@brb For this test, I needed to do what's highlighted here (disabling masquerade and BPF masquerade). Is this a regression or a known issue with net-next?

(For context, this is a new test I've added. The summary is that it's doing `(outside) host -> pod @ host A|B`, in order to validate an ingress fromCIDR + toPorts policy.)

brb replied:
Hmm, the masquerading should only influence `pod @ host A|B -> outside`, but not the opposite direction flow. Did you check what happens when the masq is enabled?

christarazi replied:
If masquerading is enabled, then what happens is the following:

```
test-output.log:{"time":"2020-07-13T20:05:10.782544267Z","verdict":"FORWARDED","ethernet":{"source":"16:de:a8:9e:9c:ee","destination":"3a:f7:49:1c:0d:66"},"IP":{"source":"192.168.36.13","destination":"10.0.1.109","ipVersion":"IPv4"},"l4":{"TCP":{"source_port":46428,"destination_port":80,"flags":{"SYN":true}}},"source":{"identity":2,"labels":["reserved:world"]},"destination":{"ID":1298,"identity":43498,"namespace":"default","labels":["k8s:io.cilium.k8s.policy.cluster=default","k8s:io.cilium.k8s.policy.serviceaccount=default","k8s:io.kubernetes.pod.namespace=default","k8s:zgroup=testDS"],"pod_name":"testds-lcz5s"},"Type":"L3_L4","node_name":"k8s2","event_type":{"type":4},"traffic_direction":"INGRESS","Summary":"TCP Flags: SYN"}
test-output.log:{"time":"2020-07-13T20:05:10.782553674Z","verdict":"FORWARDED","ethernet":{"source":"3a:f7:49:1c:0d:66","destination":"16:de:a8:9e:9c:ee"},"IP":{"source":"10.0.1.109","destination":"192.168.36.13","ipVersion":"IPv4"},"l4":{"TCP":{"source_port":80,"destination_port":46428,"flags":{"SYN":true,"ACK":true}}},"source":{"ID":1298,"identity":43498,"namespace":"default","labels":["k8s:io.cilium.k8s.policy.cluster=default","k8s:io.cilium.k8s.policy.serviceaccount=default","k8s:io.kubernetes.pod.namespace=default","k8s:zgroup=testDS"],"pod_name":"testds-lcz5s"},"destination":{"identity":2,"labels":["reserved:world"]},"Type":"L3_L4","node_name":"k8s2","reply":true,"event_type":{"type":4,"sub_type":3},"traffic_direction":"INGRESS","Summary":"TCP Flags: SYN, ACK"}
test-output.log:{"time":"2020-07-13T20:05:10.782784247Z","verdict":"FORWARDED","ethernet":{"source":"16:de:a8:9e:9c:ee","destination":"3a:f7:49:1c:0d:66"},"IP":{"source":"192.168.36.13","destination":"10.0.1.109","ipVersion":"IPv4"},"l4":{"TCP":{"source_port":46428,"destination_port":80,"flags":{"RST":true}}},"source":{"identity":2,"labels":["reserved:world"]},"destination":{"ID":1298,"identity":43498,"namespace":"default","labels":["k8s:io.cilium.k8s.policy.cluster=default","k8s:io.cilium.k8s.policy.serviceaccount=default","k8s:io.kubernetes.pod.namespace=default","k8s:zgroup=testDS"],"pod_name":"testds-lcz5s"},"Type":"L3_L4","node_name":"k8s2","event_type":{"type":4},"traffic_direction":"INGRESS","Summary":"TCP Flags: RST"}
test-output.log:{"time":"2020-07-13T20:05:11.808704554Z","verdict":"FORWARDED","ethernet":{"source":"16:de:a8:9e:9c:ee","destination":"3a:f7:49:1c:0d:66"},"IP":{"source":"192.168.36.13","destination":"10.0.1.109","ipVersion":"IPv4"},"l4":{"TCP":{"source_port":46428,"destination_port":80,"flags":{"RST":true}}},"source":{"identity":2,"labels":["reserved:world"]},"destination":{"ID":1298,"identity":43498,"namespace":"default","labels":["k8s:io.cilium.k8s.policy.cluster=default","k8s:io.cilium.k8s.policy.serviceaccount=default","k8s:io.kubernetes.pod.namespace=default","k8s:zgroup=testDS"],"pod_name":"testds-lcz5s"},"Type":"L3_L4","node_name":"k8s2","event_type":{"type":4},"traffic_direction":"INGRESS","Summary":"TCP Flags: RST"}
test-output.log:{"time":"2020-07-13T20:05:13.824950432Z","verdict":"FORWARDED","ethernet":{"source":"16:de:a8:9e:9c:ee","destination":"3a:f7:49:1c:0d:66"},"IP":{"source":"192.168.36.13","destination":"10.0.1.109","ipVersion":"IPv4"},"l4":{"TCP":{"source_port":46428,"destination_port":80,"flags":{"RST":true}}},"source":{"identity":2,"labels":["reserved:world"]},"destination":{"ID":1298,"identity":43498,"namespace":"default","labels":["k8s:io.cilium.k8s.policy.cluster=default","k8s:io.cilium.k8s.policy.serviceaccount=default","k8s:io.kubernetes.pod.namespace=default","k8s:zgroup=testDS"],"pod_name":"testds-lcz5s"},"Type":"L3_L4","node_name":"k8s2","event_type":{"type":4},"traffic_direction":"INGRESS","Summary":"TCP Flags: RST"}
```

Essentially, the following happens:

1) Outside sends TCP SYN
2) Pod responds with TCP SYN ACK
3) Outside sends TCP RST (presumably because the [2] was never received)
4) The traffic is also associated with WORLD identity when it should be associated with the identity that is created from the policy.

pchaigno replied:
> 3. Outside sends TCP RST (presumably because the [2] was never received)

More likely because the TCP SYN+ACK wasn't sent to the right address or port. Otherwise, outside would just resend a TCP SYN. Maybe SYN gets SNATed, but reply SYN+ACK doesn't get reverse SNATed?

christarazi replied:
> > 1. Outside sends TCP RST (presumably because the [2] was never received)
> 
> More likely because the TCP SYN+ACK wasn't sent to the right address or port. Otherwise, outside would just resend a TCP SYN. Maybe SYN gets SNATed, but reply SYN+ACK doesn't get reverse SNATed?

Based on the output above, I don't see how the address and port number were incorrect. Outside initiates a connection from port 46428, and all subsequent traffic uses that port. The addresses are also correct and expected.

> Maybe SYN gets SNATed, but reply SYN+ACK doesn't get reverse SNATed?

This sounds plausible to me. But I'm not familiar with this code, but it does sound like it aligns when masquerading is enabled.

pchaigno replied:
> > More likely because the TCP SYN+ACK wasn't sent to the right address or port. Otherwise, outside would just resend a TCP SYN. Maybe SYN gets SNATed, but reply SYN+ACK doesn't get reverse SNATed?
> 
> Based on the output above, I don't see how the address and port number were incorrect. Outside initiates a connection from port 46428, and all subsequent traffic uses that port. The addresses are also correct and expected.

If I understand correctly, the output above is captured at the cilium-managed node, so it may not be what the "outside" node sees. The TCP SYN is sent directly to the pod IP thanks to the route you installed on the outside node, but the reply probably goes through some SNATting in case masquerading is enabled, hence the RST. Thinking about it, it's probably expected behavior... @brb?

christarazi replied:
> > > More likely because the TCP SYN+ACK wasn't sent to the right address or port. Otherwise, outside would just resend a TCP SYN. Maybe SYN gets SNATed, but reply SYN+ACK doesn't get reverse SNATed?
> > 
> > 
> > Based on the output above, I don't see how the address and port number were incorrect. Outside initiates a connection from port 46428, and all subsequent traffic uses that port. The addresses are also correct and expected.
> 
> If I understand correctly, the output above is captured at the cilium-managed node, so it may not be what the "outside" node sees. The TCP SYN is sent directly to the pod IP thanks to the route you installed on the outside node, but the reply probably goes through some SNATting in case masquerading is enabled, hence the RST. Thinking about it, it's probably expected behavior... @brb?

Interesting. Maybe it's worth doing a tcpdump on either end so we can get a view into what's happening. Also, fwiw, this was the same behavior I was seeing on **non**-net-next kernels with either `devices` or `host-firewall` set. I imagine if I enable those on net-next, we'll still see the same RST behavior. However, what's different here on net-next **is** the fact that I need to disable masquerading and BPF masquerading.

christarazi replied:
Here's the tcpdump in the host namespace on the node of the backend pod, with masquerading enabled as well as BPF masquerading:

```
$   sudo tcpdump -n -i enp0s8 host 192.168.36.13
    tcpdump: verbose output suppressed, use -> or -v> for full protocol decode
    listening on enp0s8, link-type EN10MB (Ethernet), capture size 262144 bytes

             outside node +-----+                   +----------+ backend pod
                                v                   v
    22:39:46.174240 IP 192.168.36.13.36224 ^ 10.0.1.98.80: Flags [S], seq 51219120, win 64240, options [mss 1460,sackOK,TS val 1318196564 ecr 0,n                     op,wscale 7], length 0
+-->22:39:46.174598 IP 192.168.36.12.60468 > 192.168.36.13.36224: Flags [S.], seq 17444670, ack 51219121, win 64308, options [mss 1410,sackOK,TS                      val 1981994647 ecr 1318196564,nop,wscale 7], length 0
|   22:39:46.175214 IP 192.168.36.13.36224 > 192.168.36.12.60468: Flags [R], seq 51219121, win 0, length 0
|   22:39:47.181112 IP 192.168.36.13.36224 > 10.0.1.98.80: Flags [S], seq 51219120, win 64240, options [mss 1460,sackOK,TS val 1318197571 ecr 0,n                     op,wscale 7], length 0
|   22:39:47.181415 IP 192.168.36.12.60468 > 192.168.36.13.36224: Flags [S.], seq 33176448, ack 51219121, win 64308, options [mss 1410,sackOK,TS                      val 1981995654 ecr 1318197571,nop,wscale 7], length 0
|   22:39:47.182093 IP 192.168.36.13.36224 v 192.168.36.12.60468: Flags [R], seq 51219121, win 0, length 0
|
+---> This looks wrong for sure
```

Masquerading enabled, but BPF masquerading **disabled** (connection is successful and working):

```
22:53:01.770058 IP 192.168.36.13.36234 > 10.0.1.98.80: Flags [S], seq 568487496, win 64240, options [mss 1460,sackOK,TS val 1318992158 ecr 0,nop,wscale 7], length 0
22:53:01.770632 IP 10.0.1.98.80 > 192.168.36.13.36234: Flags [S.], seq 3972402295, ack 568487497, win 64308, options [mss 1410,sackOK,TS val 1982790243 ecr 1318992158,nop,wscale 7], length 0
22:53:01.771215 IP 192.168.36.13.36234 > 10.0.1.98.80: Flags [.], ack 1, win 502, options [nop,nop,TS val 1318992160 ecr 1982790243], length 0
22:53:01.771507 IP 192.168.36.13.36234 > 10.0.1.98.80: Flags [P.], seq 1:74, ack 1, win 502, options [nop,nop,TS val 1318992160 ecr 1982790243], length 73: HTTP: GET / HTTP/1.1
22:53:01.771620 IP 10.0.1.98.80 > 192.168.36.13.36234: Flags [.], ack 74, win 502, options [nop,nop,TS val 1982790244 ecr 1318992160], length 0
22:53:01.772725 IP 10.0.1.98.80 > 192.168.36.13.36234: Flags [P.], seq 1:577, ack 74, win 502, options [nop,nop,TS val 1982790245 ecr 1318992160], length 576: HTTP: HTTP/1.1 200 OK
22:53:01.773213 IP 192.168.36.13.36234 > 10.0.1.98.80: Flags [.], ack 577, win 501, options [nop,nop,TS val 1318992162 ecr 1982790245], length 0
22:53:01.778523 IP 192.168.36.13.36234 > 10.0.1.98.80: Flags [F.], seq 74, ack 577, win 501, options [nop,nop,TS val 1318992167 ecr 1982790245], length 0
22:53:01.779528 IP 10.0.1.98.80 > 192.168.36.13.36234: Flags [F.], seq 577, ack 75, win 502, options [nop,nop,TS val 1982790252 ecr 1318992167], length 0
22:53:01.779919 IP 192.168.36.13.36234 > 10.0.1.98.80: Flags [.], ack 578, win 501, options [nop,nop,TS val 1318992168 ecr 1982790252], length 0
```

So it looks like the culprit is definitely BPF masquerading on net-next. (It turns out that I can enable just masquerading, as long as BPF masquerading is disabled.)

joestringer replied:
We can of course continue to discuss this topic wherever you folks find easiest, but it may be worth moving this discussion to a dedicated issue to make it easier to track.

christarazi replied:
Tracked here: https://github.com/cilium/cilium/issues/12544

joestringer commented on pkg/policy/api/ingress.go:
```diff
@@ -30,12 +30,7 @@ import (
-//   then all members must match in order for the rule to take effect. The
-//   exception to this rule is the Requires field, the effects of any Requires
-//   field in any rule will apply to all other rules as well.
+// Combining FromCIDR and FromEndpoints in the same rule is not allowed.
```
Sorry, late mention but while we're fixing up this comment I think that `FromEndpoints`, `FromCIDR`, `FromCIDRSet`, `FromEntities` are mutually exclusive. I propose wording like this in an additional paragraph:

```
// - FromEndpoints, FromCIDR, FromCIDRSet and FromEntities are mutually
//   exclusive. Only one of these members may be present within an individual rule.
```

joestringer replied:
^ Can be a follow-up; cf. https://github.com/cilium/cilium/pull/12525 for egress side.

christarazi replied:
Done
